### PR TITLE
[임정현] sprint7

### PR DESCRIPTION
**Pull request by HuInDoL**
## **프로젝트 마일스톤**

- 로그 관리
- 커스텀 예외 설계
- 유효성 검사
- Actuator를 활용한 모니터링
- 단위 테스트
- 슬라이스 테스트
- 통합 테스트

### **프로파일 기반 설정 관리**

- [x]  개발, 운영 환경에 대한 프로파일을 구성하세요.
    - [x]  [ ] `application-dev.yaml`, `application-prod.yaml` 파일을 생성하세요.
    - [x]  다음과 같은 설정값을 프로파일별로 분리하세요.
        - [x]  데이터베이스 연결 정보
            
            <aside>
            📗
            
            개발은 H2
            
            운영은 DB 
            
            이런식으로?
            
            </aside>
            
        - [x]  서버 포트
            
            <aside>
            📗
            
            개발은 8080
            
            운영은 랜덤?
            
            </aside>
            

### **로그 관리**

- [x]  Lombok의 `@Slf4j` 어노테이션을 활용해 로깅을 쉽게 추가할 수 있도록 구성하세요.
- [x]  [ ] `application.yaml`에 기본 로깅 레벨을 설정하세요.
    - 기본적으로 `info` 레벨로 설정합니다.
- [x]  환경 별 적절한 로깅 레벨을 프로파일 별로 설정해보세요.
    - SQL 로그를 보기위해 설정했던 레벨은 유지합니다.
    - 우리가 작성한 프로젝트의 로그는 개발 환경에서 `debug`, 운영 환경에서는 `info` 레벨로 설정합니다.
- [x]  Spring Boot의 기본 로깅 구현체인 Logback의 설정 파일을 구성하세요.
    
    (교안 Logback 참조)
    
    - [x]  `logback-spring.xml` 파일을 생성하세요.
    - [x]  다음 예시와 같은 로그 메시지를 출력하기 위한 로깅 패턴과 출력 방식을 커스터마이징하세요.
        - 로그 출력 예시
            
            ```markdown
            
            # 패턴
            {년}-{월}-{일} {시}:{분}:{초}:{밀리초} [{스레드명}] {로그 레벨(5글자로 맞춤)} {로거 이름(최대 36글자)} - {로그 메시지}{줄바꿈}
            
            # 예시
            25-01-01 10:33:55.740 [main] DEBUG c.s.m.discodeit.DiscodeitApplication - Running with Spring Boot v3.4.0, Spring v6.2.0
            
            ```
            
    - [x]  콘솔과 **파일**에 동시에 로그를 기록하도록 설정하세요.
        - [x]  파일은 `{프로젝트 루트}/.logs` 경로에 저장되도록 설정하세요.
    - [x]  로그 파일은 **일자별로 롤링**되도록 구성하세요.
    - [x]  로그 파일은 **30일**간 보관하도록 구성하세요.
- [x]  서비스 레이어와 컨트롤러 레이어의 주요 메소드에 로깅을 추가하세요.
    - [x]  로깅 레벨을 적절히 사용하세요: ERROR, WARN, INFO, DEBUG
    - [x]  다음과 같은 메소드에 로깅을 추가하세요:
        - [x]  사용자 생성/수정/삭제
        - [x]  채널 생성/수정/삭제
        - [x]  메시지 생성/수정/삭제
        - [x]  파일 업로드/다운로드

### **예외 처리 고도화**

- [x]  **커스텀 예외**를 설계하고 구현하세요.
    - 패키지명: `com.sprint.mission.discodeit.exception[.{도메인}]`
    - [x]  `ErrorCode` Enum 클래스를 통해 예외 코드명과 메시지를 정의하세요.
        - 아래는 예시입니다. 필요하다고 판단되는 다양한 코드를 정의하세요.
        - 예시
            
            [[6ag3lzl9i-image.png](https://bakey-api.codeit.kr/api/files/resource?root=static&seqId=13906&version=1&directory=/6ag3lzl9i-image.png&name=6ag3lzl9i-image.png)](https://bakey-api.codeit.kr/api/files/resource?root=static&seqId=13906&version=1&directory=/6ag3lzl9i-image.png&name=6ag3lzl9i-image.png)
            
    - [x]  모든 예외의 기본이 되는 `DiscodeitException` 클래스를 정의하세요.
        - 클래스 다이어그램
            
            [[j5vtp941a-image.png](https://bakey-api.codeit.kr/api/files/resource?root=static&seqId=13906&version=1&directory=/j5vtp941a-image.png&name=j5vtp941a-image.png)](https://bakey-api.codeit.kr/api/files/resource?root=static&seqId=13906&version=1&directory=/j5vtp941a-image.png&name=j5vtp941a-image.png)
            
        - `details`는 예외 발생 상황에 대한 추가정보를 저장하기 위한 속성입니다.
            - 예시
                - 조회 시도한 사용자의 ID 정보
                - 업데이트 시도한 PRIVATE 채널의 ID 정보
    - [x]  `DiscodeitException`을 상속하는 주요 도메인 별 메인 예외 클래스를 정의하세요.
        - `UserException`, `ChannelException` 등
        - 실제로 활용되는 클래스라기보다는 **예외 클래스의 계층 구조를 명확하게 하기 위한 클래스** 입니다.
    - [x]  도메인 메인 예외 클래스를 상속하는 구체적인 예외 클래스를 정의하세요.
        - `UserNotFoundException`, `UserAlreadyExistException` 등 필요한 예외를 정의하세요.
        - 예시
            
            [[a6f585icy-image.png](https://bakey-api.codeit.kr/api/files/resource?root=static&seqId=13906&version=1&directory=/a6f585icy-image.png&name=a6f585icy-image.png)](https://bakey-api.codeit.kr/api/files/resource?root=static&seqId=13906&version=1&directory=/a6f585icy-image.png&name=a6f585icy-image.png)
            
- [x]  기존에 구현했던 예외를 **커스텀 예외**로 대체하세요.
    - `NoSuchElementException`
    - `IllegalArgumentException`
    - …
- [x]  `ErrorResponse`를 통해 **일관된 예외 응답**을 정의하세요.
    - 클래스 다이어그램
        
        [[3gqeampkw-image.png](https://bakey-api.codeit.kr/api/files/resource?root=static&seqId=13906&version=1&directory=/3gqeampkw-image.png&name=3gqeampkw-image.png)](https://bakey-api.codeit.kr/api/files/resource?root=static&seqId=13906&version=1&directory=/3gqeampkw-image.png&name=3gqeampkw-image.png)
        
    - `int status`: HTTP 상태코드
    - `String exceptionType`: 발생한 예외의 클래스 이름
- [x]  앞서 정의한 `ErrorResponse`와 `@RestControllerAdvice`를 활용해 예외를 처리하는 예외 핸들러를 구현하세요.
    - 모든 핸들러는 일관된 응답(`ErrorResponse`)을 가져야 합니다.

### **유효성 검사**

- [x]  Spring Validation 의존성을 추가하세요.
- [x]  주요 Request DTO에 **제약 조건 관련** 어노테이션을 추구하세요.
    - `@NotNull`, `@NotBlank`, `@Size`, `@Email` 등
- [x]  컨트롤러에 `@Valid` 를 사용해 요청 데이터를 검증하세요.
- [x]  검증 실패 시 발생하는 `MethodArgumentNotValidException`을 **전역 예외 핸들러**에서 처리하세요.
- [x]  유효성 검증 실패 시 상세한 오류 메시지를 포함한 **응답**을 반환하세요.

### **Actuator**

- [x]  Spring Boot Actuator 의존성을 추가하세요.
- [x]  기본 Actuator 엔드포인트를 설정하세요.
    - health, info, metrics, loggers
- [x]  Actuator info를 위한 **애플리케이션 정보**를 추가하세요.
    - 애플리케이션 이름: `Discodeit`
    - 애플리케이션 버전: `1.7.0`
    - 자바 버전: `17`
    - 스프링 부트 버전: `3.4.0`
    - 주요 설정 정보
        - 데이터소스: url, 드라이버 클래스 이름
        - jpa: ddl-auto
        - storage 설정: type, path
        - multipart 설정: max-file-size, max-request-size
- [ ]  Spring Boot 서버를 실행 후 각종 정보를 확인해보세요.
    - [ ]  `/actuator/info`
    - [x]  `/actuator/metrics`
    - [x]  `/actuator/health`
    - [x]  `/actuator/loggers`
    

### **단위 테스트**

- [x]  **서비스 레이어**의 주요 메소드에 대한 **단위 테스트**를 작성하세요.
    - [x]  다음 서비스의 핵심 메소드에 대해 **각각 최소 2개 이상(성공, 실패)**의 테스트 케이스를 작성하세요.
        - [x]  UserService: create, update, delete 메소드
        - [x]  ChannelService: create(PUBLIC, PRIVATE), update, delete, findByUserId 메소드
        - [x]  MessageService: create, update, delete, findByChannelId 메소드
    - [x]  [ ] `Mockito`를 활용해 Repository 의존성을 모의(mock)하세요.
    - [x]  [ ] `BDDMockito`를 활용해 테스트 가독성을 높이세요.

### **슬라이스 테스트**

- [x]  **레포지토리 레이어**의 **슬라이스 테스트**를 작성하세요.
    
    (레포지토리 슬라이스 테스트는 며칠을 해도 안돼서 포기)
    
    - [x]  [ ] `@DataJpaTest`를 활용해 테스트를 구현하세요.
    - [x]  테스트 환경을 구성하는 프로파일을 구성하세요.
        - [x]  [ ] `application-test.yaml`을 생성하세요.
        - [x]  데이터소스는 **H2 인메모리 데이터 베이스**를 사용하고, **PostgreSQL 호환 모드**로 설정하세요.
        - [x]  H2 데이터베이스를 위해 필요한 **의존성**을 추가하세요.
        - [x]  테스트 시작 시 **스키마를 새로 생성**하도록 설정하세요.
        - [x]  **디버깅**에 용이하도록 로그 레벨을 적절히 설정하세요.
    - [x]  테스트 실행 간 `test` **프로파일**을 활성화 하세요.
    - [x]  JPA Audit 기능을 활성화 하기 위해 테스트 클래스에 `@EnableJpaAuditing`을 추가하세요.
    - [x]  주요 레포지토리(User, Channel, Message)의 주요 쿼리 메소드에 대해 각각 최소 2개 이상(성공, 실패)의 테스트 케이스를 작성하세요.
        - [x]  **커스텀 쿼리** 메소드
        - [x]  **페이징 및 정렬** 메소드
- [x]  **컨트롤러 레이어**의 **슬라이스 테스트**를 작성하세요.
    - [x]  `@WebMvcTest`를 활용해 테스트를 구현하세요.
    - [x]  `WebMvcTest`에서 자동으로 등록되지 않는 유형의 Bean이 필요하다면 `@Import`를 활용해 추가하세요.
        - 예시
            
            ```java
            
            @Import({ErrorCodeStatusMapper.class})
            
            ```
            
    - [x]  주요 컨트롤러(User, Channel, Message)에 대해 최소 2개 이상(성공, 실패)의 테스트 케이스를 작성하세요.
    - [x]  **MockMvc**를 활용해 컨트롤러를 테스트하세요.
    - [x]  서비스 레이어를 **모의(mock)**하여 컨트롤러 로직**만** 테스트하세요.
    - [x]  **JSON 응답을 검증**하는 테스트를 포함하세요.

### **통합 테스트**

- [x]  통합 테스트 환경을 구성하세요.
    - [x]  [ ] `@SpringBootTest`를 활용해 Spring 애플리케이션 컨텍스트를 로드하세요.
    - [x]  **H2 인메모리 데이터베이스**를 활용하세요.
    - [x]  **테스트용 프로파일**을 구성하세요.
- [x]  주요 API 엔드포인트에 대한 통합 테스트를 작성하세요.
    - [x]  주요 API에 대해 최소 2개 이상의 테스트 케이스를 작성하세요.
        - [x]  사용자 관련 API (생성, 수정, 삭제, 목록 조회)
        - [x]  채널 관련 API (생성, 수정, 삭제)
        - [x]  메시지 관련 API (생성, 수정, 삭제, 목록 조회)
    - [x]  각 테스트는 `@Transactional`을 활용해 독립적으로 실행하세요.

### **MDC를 활용한 로깅 고도화**

- [x]  요청 ID, 요청 URL, 요청 방식 등의 정보를 **MDC**에 추가하는 인터셉터를 구현하세요.
    - [x]  클래스명: `MDCLoggingInterceptor`
    - [x]  패키지명: `com.**.discodeit.config`
    - [x]  요청 ID는 랜덤한 문자열로 생성합니다. (**UUID**)
    - [x]  요청 ID는 **응답 헤더에 포함**시켜 더 많은 분석이 가능하도록 합니다.
        - 헤더 이름: `Discodeit-Request-ID`
- [x]  [ ] `WebMvcConfigurer`를 통해 `MDCLoggingInterceptor`를 등록하세요.
    - [x]  클래스명: `WebMvcConfig`
    - [x]  패키지명: `com.**.discodeit.config`
- [x]  Logback 패턴에 MDC 값을 포함시키세요.
    - 로그 출력 예시
        
        ```markdown
        
        # 패턴
        {년}-{월}-{일} {시}:{분}:{초}:{밀리초} [{스레드명}] {로그 레벨(5글자로 맞춤)} {로거 이름(최대 36글자)} [{MDC:요청ID} | {MDC:요청 메소드} | {MDC:요청 URL}] - {로그 메시지}{줄바꿈}
        
        # 예시
        25-01-01 10:33:55.740 [main] DEBUG o.s.api.AbstractOpenApiResource [827cbc0b | GET | /v3/api-docs] - Init duration for springdoc-openapi is: 216 ms
        
        ```
        

### **Spring Boot Admin을 활용한 메트릭 가시화**

- [x]  [Spring Boot Admin 서버](https://www.notion.so/Spring-Boot-Admin-2178d8ad92078040a5cbc23968b3afa1?pvs=21)를 구현할 모듈을 생성하세요.
    
    [Spring Boot Admin 서버](https://www.notion.so/Spring-Boot-Admin-2178d8ad92078040a5cbc23968b3afa1?pvs=21)
    
    - IntelliJ 화면 참고
        
        [[9l6b0q2dn-image.png](https://bakey-api.codeit.kr/api/files/resource?root=static&seqId=13907&version=1&directory=/9l6b0q2dn-image.png&name=9l6b0q2dn-image.png)](https://bakey-api.codeit.kr/api/files/resource?root=static&seqId=13907&version=1&directory=/9l6b0q2dn-image.png&name=9l6b0q2dn-image.png)
        
    - 모듈 정보는 다음과 같습니다.
        
        [[b8812d4su-image.png](https://bakey-api.codeit.kr/api/files/resource?root=static&seqId=13907&version=1&directory=/b8812d4su-image.png&name=b8812d4su-image.png)](https://bakey-api.codeit.kr/api/files/resource?root=static&seqId=13907&version=1&directory=/b8812d4su-image.png&name=b8812d4su-image.png)
        
    - 의존성
        
        [[zjh3frl0m-image.png](https://bakey-api.codeit.kr/api/files/resource?root=static&seqId=13907&version=1&directory=/zjh3frl0m-image.png&name=zjh3frl0m-image.png)](https://bakey-api.codeit.kr/api/files/resource?root=static&seqId=13907&version=1&directory=/zjh3frl0m-image.png&name=zjh3frl0m-image.png)
        
- [x]  `admin` 모듈의 메인 클래스에 `@EnableAdminServer` 어노테이션을 추가하고, 서버는 9090번 포트로 설정합니다.
    
    ```java
    
    import de.codecentric.boot.admin.server.config.EnableAdminServer;
    
    @SpringBootApplication
    @EnableAdminServer
    public class AdminApplication {
        public static void main(String[] args) {
            SpringApplication.run(AdminApplication.class, args);
        }
    }
    
    ```
    
    ```yaml
    
    # application.yaml
    spring:
        application:
            name: admin
    server:
        port: 9090
    
    ```
    
- [x]  `admin` 서버 실행 후 [[localhost:9090/applications](http://localhost:9090/applications)](http://localhost:9090/applications) 에 접속해봅니다.
- [x]  discodeit 프로젝트에 Spring Boot Admin Client를 적용합니다.
    - [x]  의존성을 추가합니다.
        
        ```groovy
        
        dependencies {
            ...
            implementation 'de.codecentric:spring-boot-admin-starter-client:3.4.5
        }
        
        ```
        
    - [x]  admin 서버에 등록될 수 있도록 설정 정보를 추가합니다.
        
        ```yaml
        
        # application.yml
        spring:
          application:
            name: discodeit
            ...
          boot:
            admin:
              client:
                instance:
                  name: discodeit
        ...
        
        ```
        
        ```yaml
        
        # application-dev.yml
        spring:
          application:
            name: discodeit
            ...
          boot:
            admin:
              client:
                url: http://localhost:9090
        ...
        
        ```
        
        ```yaml
        
        # application-prod.yml
        spring:
          application:
            name: discodeit
            ...
          boot:
            admin:
              client:
                url: ${SPRING_BOOT_ADMIN_CLIENT_URL}
        ...
        
        ```
        
    - [ ]  discodeit 서버를 실행하고, admin 대시보드에 discodeit 인스턴스가 추가되었는지 확인합니다.
        
        → 그대로 따라했는데 안됨. 계속 고쳐봐도 안됨……
        
- [ ]  admin 대시보드 화면을 조작해보면서 각종 메트릭 정보를 확인해보세요.
    - 주요 API의 요청 횟수, 응답시간 등
    - 서비스 정보

### **테스트 커버리지 관리**

- [x]  JaCoCo 플러그인을 추가하세요.
    
    ```
    
    plugins {
        id 'jacoco'
    }
    
    test {
        finalizedBy jacocoTestReport
    }
    
    jacocoTestReport {
        dependsOn test
        reports {
            xml.required = true
            html.required = true
        }
    }
    
    ```
    
- [x]  테스트 실행 후 생성된 리포트를 분석해보세요.
    - 리포트는 `build/reports/jacoco` 경로에서 확인할 수 있습니다.
- [ ]  `com.sprint.mission.discodeit.service.basic` 패키지에 대해서 **60%** 이상의 코드 커버리지를 달성하세요.

1. 테스트코드를 어떻게 작성해야 할지 모르겠습니다. 일단 고민하다가 한개씩은 썼는데, 너무 시간이 많이 들 것 같아서, 대부분 생성형 AI의 도움을 받았습니다.
    1. 어떤 것을 검증해야 하는지 
    2. 그 어떤 것을 어떤 식으로 검증해야 하는지
    
    → 위 두가지에 대해서 참고할 만한 자료가 있다면, 저에게 일러주시면 정말 감사하겠습니다.
    
2. [`localhost:8080/actuator/info`를](http://[localhost:8080/actuator/info를](http://localhost:8080/actuator/info%EB%A5%BC)) 들어가면 `[]`  만 나옴. 근데 이걸 야믈파일을 아무리 고쳐도 빈파일만 나옵니다.
image

3. admin 서버 연결을 하려고 모듈을 추가했는데, discodeit 인스턴스가 admin server에 인식이 되지 않습니다. 수정하려고 했지만, ai 힘을 빌리려다가 계속 DB를 삭제하는 등의 오류가 다수 발생하고 모듈이 제대로 삭제되지 않는 등의 버그가 발생해서 시간이 오래 끌렸습니다. 그래서 일단 포기했습니다. 설정파일에서 어떤 것을 바로잡아야 할지 모르겠습니다.

image

### CODE REVIEW

**Comment by IDontHaveBrain** [comment]
정현님, XXXException 파일들이 누락된거 같습니다!
빌드 가능하게 현재 PR에 XXXException 파일까지 추가 commit & push하여 주세요!

2.
management.info.env.enabled 항목을 true로 설정하여 주시면 됩니다.
기본적인 서칭이나 GPT 질문으로 해결이 안되거나, GPT가 수정하여 준 내용을 이해 못할경우.
라이브러리 관련 이슈 및 수정은 보통 해당 라이브러리 공식 문서를 보면서 구성 및 수정하는게 가장 Best라고 생각합니다.
https://github.com/codecentric/spring-boot-admin
https://docs.spring-boot-admin.com/3.4.7/docs/installation-and-setup/

spring-boot-admin github에서 README.md 문서에 적힌 Compatibility Matrix 항목도 꼭 확인 바랍니다.
확인 하셨다면 build.gradle 에서 수정해야 할 버전이 있을거라 생각됩니다.

```yaml
management:
  endpoints:
    web:
      exposure:
        include: health, info, metrics, loggers
      base-path: /actuator
  endpoint:
    health:
      # 상태 상세 정보 출력 여부 (never, when-authorized, always)
      # 추후 never로 수정
      show-details: never
      show-components: never
  info:
    env:
      enabled: true
```

3.
제 환경에서는 인식이 되어서, 정현님이 구성하신 admin server 까지 git 및 pr에 추가하여 주세요.
그리고 admin server 구동, discodeit 서버 구동. 동시에 두개의 서버를 실행하여 두신게 맞을까요 ?

File: src/main/java/com/sprint/mission/discodeit/config/MDCLoggingInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                    // 다음 필터 체인으로 요청 전달
+                    filterChain.doFilter(request, response);
+                } finally {
+                    // 요청 처리 완료 후 MDC 정리 (메모리 누수 방지)
+                    MDC.clear();
+                }
+            }
+        };
+    }
+
```
Comment (IDontHaveBrain) [inline src/main/java/com/sprint/mission/discodeit/config/MDCLoggingInterceptor.java]: Filter - Spring 컨텍스트의 DispatcherServlet 앞에서 동작 -> 요청의 가장 바깥 단에서 처리해야 하는 로깅, 인코딩 등에 적합
Interceptor - DispatcherServlet과 컨트롤러 사이에서 동작 ->사용자 인증, 권한 체크 등 비즈니스 로직에 더 가까운 처리에 적합

MDC 설정은 요청 전체에 걸쳐 유효해야 하므로, 가장 먼저 실행되는 Filter로 구현하는 것이 더 일반적입니다.

Filter와 Interceptor는 다릅니다!
implements HandlerInterceptor 제거 및 WebMvcConfig에 필터 제거해도 동일하게 동작 할 것으로 보입니다!

File: src/main/java/com/sprint/mission/discodeit/exception/GlobalExceptionHandler.java [hunk 2]
```diff
@@ -1,36 +1,136 @@
...
-public class GlobalExceptionHandler extends RuntimeException /* 비검사 예외 */ {
+public class GlobalExceptionHandler extends RuntimeException {
 
-    @ExceptionHandler(InvalidArgException.class)
-    protected ResponseEntity<ErrorDto> handleInvalidArgException(InvalidArgException e) {
-        return ErrorDto.toResponseEntity(e.getErrorCode());
+    /**
+     * User 도메인 예외를 처리합니다.
+     */
+    @ExceptionHandler(UserException.class)
```
Comment (IDontHaveBrain) [inline src/main/java/com/sprint/mission/discodeit/exception/GlobalExceptionHandler.java:47]: 다들 동일하신게 아마 미션에서 추구하는 코드 방향 자체가 도메인별 커스텀예외를 각각 정의하는 듯 하지만, 개인적인 의견은 아래와 같습니다.

DiscodeitException 을 제외한 XXXException 들은 오버엔지니어링 내지 불필요하여 보입니다.

대부분의 커스텀 예외 핸들러(handleXXXException)의 내부 로직이 거의 동일합니다.
새로운 로직과 기능이 추가되어 새로운 예외 타입을 추가할 때마다 커스텀 예외 클래스를 계속 추가해야 하며 GlobalExceptionHandler 또한 수정이 필요하므로 매우 번거로울 것으로 보입니다.

현재 코드베이스에선 DiscodeitException 하나만으로 충분하지 않을까요 ?
커스텀 메세지는 생성자나 팩토리메서드를 통해 매개변수로 받으셔도 되고, ErrorCode 상 메세지에 플레이스 홀더("{}는 이미 존재하는 사용자입니다.", "{user}는 이미 존재하는 사용자입니다.")를 사용하여 치환하여도 됩니다.

아니면 도메인별 커스텀 예외 클래스는 유지 하면서 최소한 핸들러를 그때그때 등록 할 필요는 없이 아래와 같이 공통 부모가 되는 Exception만 핸들링 해두는게 편할 것 같습니다.
```java
    @ExceptionHandler(DiscodeitException.class)
    public ResponseEntity<ErrorResponse> handleDiscodeitException(DiscodeitException ex) {
        ErrorCode errorCode = ex.getErrorCode();
        ErrorResponse response = new ErrorResponse(
            errorCode.getStatus().value(),
            errorCode.getCode(),
            ex.getMessage(), // 각 예외에 맞는 메시지를 사용
            ex.getTimestamp(),
            ex.getDetails()
        );
        return ResponseEntity.status(errorCode.getStatus()).body(response);
    }
```

File: src/main/java/com/sprint/mission/discodeit/mapper/ChannelMapper.java [hunk 3]
```diff
@@ -1,46 +1,57 @@
...
+
+    protected Instant resolveLastMessageAt(Channel channel) {
+        try {
+            return messageRepository.findTop1ByChannelIdOrderByCreatedAtDesc(channel.getId())
+                    .map(message -> message.getCreatedAt())
+                    .orElse(null);
+        } catch (Exception e) {
+            return null;
+        }
     }
```
Comment (IDontHaveBrain) [inline src/main/java/com/sprint/mission/discodeit/mapper/ChannelMapper.java:56]: 현재 아키텍처 레이어를 크게 4가지로 나누어서 보면 Controller, Service, Repository, Mapper 입니다.

레이어를 나누는 이유는 뭘까요?
Mapper, Service 각각에 역할 내지 책임은 무엇일까요 ?
Mapper에서 Repository 를 주입받아서 사용하는 상황에서 Service 레이어와의 차이점은 무엇일까요 ?

File: src/main/java/com/sprint/mission/discodeit/service/basic/BasicReadStatusService.java [hunk 4]
```diff
@@ -1,103 +1,149 @@
...
+                    return new ReadStatusNotFoundException("채널을 찾을 수 없습니다.");
+                });
 
-        if (!userRepository.existsById(userId)) {
-            throw new NoSuchElementException("User with id " + userId + " does not exist");
-        }
-        if (!channelRepository.existsById(channelId)) {
-            throw new NoSuchElementException("Channel with id " + channelId + " does not exist");
-        }
         if (readStatusRepository.findAllByUserId(userId).stream()
```
Comment (IDontHaveBrain) [inline src/main/java/com/sprint/mission/discodeit/service/basic/BasicReadStatusService.java:65]: ReadStatusRepository 레포지토리에 existsByUserIdAndChannelId 를 사용 안하신 이유가 있을까요 ?

통계나 너무 복잡한 조건이 아니라면 대부분 DB단에서 필터링하여 조회 하는게 성능, 메모리적으로 유리합니다.(특히 데이터가 많을수록 이 경우엔 사용자가 참여한 채널이 많을수록)

**Review by IDontHaveBrain — APPROVED** [review]
1. 
참고용 글 - https://devbksheen.tistory.com/entry/%ED%85%8C%EC%8A%A4%ED%8A%B8-%EC%BD%94%EB%93%9C-%EC%9E%91%EC%84%B1%ED%95%98%EA%B8%B0JUnit-TDD-BDD-Mockito

이건 모든 개발자가 항상 하는 고민이고, 정답이 하나만 있는 것은 아닙니다.

무엇을 검증해야 하는가? (What to test?)
테스트의 목표는 "내 코드가 의도한 대로 동작하는 것을 보장"하는 것입니다.
이걸 기준으로 생각해보면 좋습니다.

Happy Path (성공 시나리오): 가장 기본적인 기능이 정상적으로 동작하는 경우. 예를 들어, UserService.create에서 유효한 정보로 유저가 성공적으로 생성되고, 결과 DTO가 올바르게 반환되는지 검증합니다.
Edge Cases (경계값/엣지 케이스): 로직의 분기점이 되는 경계 조건을 테스트합니다.
null이나 빈 문자열("")이 입력되었을 때
목록이 비어있거나(List.isEmpty()) 하나만 있을 때
숫자가 0, -1일 때
Unhappy Path (실패/예외 시나리오): 의도적으로 잘못된 상황을 만들어 예외가 발생하는지 검증합니다.
UserService.create에서 중복된 이메일로 가입 시도 시 DuplicatedUserException이 발생하는가?
존재하지 않는 ID로 조회를 시도할 때 UserNotFoundException이 발생하는가?
권한 없는 유저가 수정을 시도할 때 PermissionDeniedException (가칭)이 발생하는가?
Side Effects (부수 효과): 특정 메소드가 호출되었을 때, 다른 객체나 시스템에 올바른 영향을 미쳤는지 검증합니다.
UserService.delete(userId) 호출 시, userRepository.deleteById(userId)가 정말로 호출되었는가?
프로필 이미지와 함께 유저 생성 시, binaryContentStorage.put() 메소드가 호출되었는가?


어떻게 검증해야 하는가? (How to test?)
검증 대상에 따라 사용하는 방법이 다릅니다.

값(Value) 검증 - Assertions 사용: 메소드의 반환값이 기대하는 값과 일치하는지 확인합니다.
assertEquals(expected, actual): 두 값이 같은지 비교.
assertNotNull(object): 객체가 null이 아닌지 확인.
assertTrue(boolean): 조건이 참인지 확인.
예외(Exception) 검증 - assertThrows 사용: 특정 상황에서 의도한 예외가 발생하는지 확인합니다.
assertThrows(ExpectedException.class, () -> { /* 예외를 발생시키는 코드 */ });
상호작용(Interaction) 검증 - Mockito.verify (또는 BDDMockito then) 사용: Mock 객체의 특정 메소드가 예상대로 호출되었는지 확인합니다. (주로 반환값이 void인 메소드나, 부수 효과를 검증할 때 사용)
then(mockObject).should().methodName(arguments);: methodName이 호출되었는지 검증.
then(mockObject).should(never()).methodName(any());: methodName이 호출되지 않았는지 검증.
then(mockObject).should(times(2)).methodName();: 2번 호출되었는지 검증.

ex)
```java
@Test
@DisplayName("[성공] 사용자가 정상적으로 생성된다")
void createUser_Success() {
    // given (준비)
    UserCreateRequest request = new UserCreateRequest("test", "test@a.com", "pass");
    User savedUser = new User("test", "test@a.com", "pass", null);
    // ... Mock 객체 설정 ...
    given(userRepository.existsByEmail(anyString())).willReturn(false);
    given(userRepository.saveAndFlush(any(User.class))).willReturn(savedUser);
    given(userMapper.toDto(any(User.class))).willReturn(new UserDto(...));

    // when (실행)
    UserDto result = userService.create(request, Optional.empty());

    // then (검증)
    // 1. 값 검증
    assertNotNull(result);
    assertEquals("test", result.username());
    
    // 2. 상호작용 검증
    then(userRepository).should().existsByEmail("test@a.com"); // 이메일 중복 체크를 했는가?
    then(userRepository).should().saveAndFlush(any(User.class)); // 저장을 했는가?
    then(userStatusRepository).should().save(any(UserStatus.class)); // 상태 정보도 저장했는가?
}

@Test
@DisplayName("[실패] 중복된 이메일로 생성 시 DuplicatedUserException 발생")
void createUser_Fail_DuplicatedEmail() {
    // given
    UserCreateRequest request = new UserCreateRequest("test", "test@a.com", "pass");
    given(userRepository.existsByEmail("test@a.com")).willReturn(true);
    
    // when & then
    // 3. 예외 검증
    assertThrows(DuplicatedUserException.class, () -> {
        userService.create(request, Optional.empty());
    });
    
    // 4. 상호작용 검증 (실패했으므로 저장은 안 되어야 함)
    then(userRepository).should(never()).saveAndFlush(any());
}
```

이번 리뷰가 정현님의 성장에 도움이 되었으면 좋겠습니다, 수고하셨습니다!

**Comment by HuInDoL** [comment]
안녕하세요 멘토님, 임정현입니다! 늦게 확인해서 죄송합니다. 늦은 시간에도 정성스러운 리뷰 정말 감사드립니다! 확인 후 답변 드리겠습니다! 
일단 리뷰해주신 내용은 아직 반영 하지 않고, 누락된 파일들만 커밋 후 푸쉬했습니다. 
시간 되실 때 확인 부탁드립니다!

File: src/main/java/com/sprint/mission/discodeit/exception/GlobalExceptionHandler.java (hunk 2 above)
Comment (HuInDoL) [reply src/main/java/com/sprint/mission/discodeit/exception/GlobalExceptionHandler.java:47]: 미션8 완성하고, 제출하기 전 시간적 여유가 있을 때 리뷰해주신 내용 반영토록 하겠습니다!

File: src/main/java/com/sprint/mission/discodeit/service/basic/BasicReadStatusService.java (hunk 4 above)
Comment (HuInDoL) [reply src/main/java/com/sprint/mission/discodeit/service/basic/BasicReadStatusService.java:65]: 희미한 기억으로는 구현해만 초점을 맞춰서 계속 변경하다가, 미션 7 구현하면서 로직 수정할 때  존재 여부보다 find를 써야할 것 같아서 exists 로직을 삭제한 것 같습니다. 
미처 생각을 못했는데 말씀주신대로 DB 쿼리로 필터링하는게 확실히 유리한 것 같습니다. 
일단 미션 진행하고 리팩토링 할 수 있을 때 반영해보겠습니다

File: src/main/java/com/sprint/mission/discodeit/config/MDCLoggingInterceptor.java (hunk 1 above)
Comment (HuInDoL) [reply src/main/java/com/sprint/mission/discodeit/config/MDCLoggingInterceptor.java]: 리뷰주신대로 MDCLoggingFilter 로 바꾸고 implements도 삭제하고 WebMvcConfig도 삭제했는데, 잘 동작합니다.
둘의 구분에 대해 알려주셔서 감사합니다.
